pkg/progress/progress.pipeline.go

```diff
@@ -0,0 +1,157 @@
+package progress
+
+import (
+	"fmt"
+	"time"
+
+	"bunnyshell.com/sdk"
+	"github.com/briandowns/spinner"
+)
+
+type PipelineSyncer func() (*sdk.PipelineItem, error)
+
+type Progress struct {
+	Options Options
+
+	spinner *spinner.Spinner
+
+	stages map[string]bool
+}
+
+type Options struct {
+	Interval time.Duration
+}
+
+func NewOptions() *Options {
+	return &Options{
+		Interval: defaultUpdate,
+	}
+}
+
+func NewPipeline(options Options) *Progress {
+	spinner := newSpinner()
+	spinner.Prefix = fmt.Sprintf(
+		"%s Fetching pipeline status... ",
+		prefixWait,
+	)
+
+	return &Progress{
+		Options: options,
+
+		spinner: spinner,
+		stages:  map[string]bool{},
+	}
+}
+
+func (p *Progress) Update(pipelineSync PipelineSyncer) error {
+	for {
+		pipeline, err := pipelineSync()
+		if err != nil {
+			return err
+		}
+
+		if pipeline == nil {
+			return nil
+		}
+
+		if !p.UpdatePipeline(pipeline) {
+			return nil
+		}
+
+		time.Sleep(p.Options.Interval)
+	}
+}
+
+func (p *Progress) UpdatePipeline(pipeline *sdk.PipelineItem) InProgress {
+	if pipeline == nil {
+		return false
+	}
+
+	p.spinner.Prefix = "Processing Pipeline "
+
+	for _, stage := range pipeline.GetStages() {
+		switch p.setStage(stage) {
+		case Success:
+			continue
+		case Failed:
+			return false
+		case Synced:
```

Comment:
I think this needs some default that errors out / breaks the loop - just in case someone adds a new `UpdateStatus` without touching this code.